docs/configuration/http_filters/lua_filter.rst

```diff
@@ -0,0 +1,325 @@
+supported Lua version is mostly 5.1 with some 5.2 features. See the `LuaJIT documentation
+<http://luajit.org/extensions.html>`_ for more details.
+
+The filter only supports loading Lua code in-line in the configuration. If local filesystem code
+is desired, a trivial in-line script can be used to load the rest of the code from the local
+environment.
+
+The design of the filter and Lua support at a high level is as follows:
+
+* All Lua environments are :ref:`per worker thread <arch_overview_threading>`. This means that
+  there is no truly global data. Any globals create and populated at load time will be visible
+  from each worker thread in isolation. True global support may be added via an API in the future.
+* All scripts are run as coroutines. This means that they are written in a synchronous style even
+  though they may perform complex asynchronous tasks. This makes the scripts substantially easier
+  to write. All network/async processing is performed by Envoy via a set of APIs. Envoy will
+  yield the script as appropriate and resume it when async tasks are complete.
+* **Do not perform blocking operations from scripts.** It is critical for performance that
+  Envoy APIs are used for all IO.
+
+Currently supported high level features
+---------------------------------------
+
+**NOTE:** It is expected that this list will expand over time as the filter is used in production.
+The API surface has been kept small on purpose. The goal is to make scripts extremely simple and
+safe to write. Very complex or high performance use cases are assumed to use the native C++ filter
+API.
+
+* Inspection of headers, body, and trailers while streaming in either the request flow, response
+  flow, or both.
+* Modification of headers and trailers.
+* Blocking and buffering the full request/response body for inspection.
+* Performing an outbound async HTTP call to an upstream host. Such a call can be performed while
+  buffering body data so that when the call completes upstream headers can be modified.
+* Performing a direct response and skipping further filter iteration. For example, a script
+  could make an upstream HTTP call for authentication, and then directly respond with a 403
+  response code.
+
+Configuration
+-------------
+
+.. code-block:: json
+
+  {
+    "name": "lua",
+    "config": {
+      "inline_code": "..."
+    }
+  }
+
+inline_code
+  *(required, string)* The Lua code that Envoy will execute. This can be a very small script that
+  further loads code from disk if desired. Note that if JSON configuration is used, the code must
+  be properly escaped. YAML configuration may be easier to read since YAML supports multi-line
+  strings so complex scripts can be easily expressed inline in the configuration.
+
+Script examples
+---------------
+
+.. code-block:: lua
+
+  -- Called on the request path.
+  function envoy_on_request(request_handle)
```

Comment:
If the name of this function is fixed, it would be clearer if this requirement is specified above (Envoy will call a predefined set of callbacks in request and response path).

Reply:
Will reorganize slightly to make this more clear.